File: invoice_me/frontend/lib/types/payment-information.ts
import { AcceptedPaymentMethods } from "../enums/invoicing";

export interface BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods;
  enabled: boolean;
  information_present: boolean;
  display_name?: string;
  description?: string;
}

// Specific payment method interfaces
export interface EFTPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.EFT;
  bank_name: string;
  account_holder: string;
  account_number: string;
  branch_code: string;
  swift_code?: string;
}

export interface InstantEFTPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.INSTANT_EFT;
  bank_name: string;
  account_holder: string;
  account_number: string;
  branch_code: string;
}

export interface SnapScanPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.SNAPSCAN;
  merchant_id: string;
  store_id?: string;
  qr_code_url?: string;
}

export interface PayShapPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.PAYSHAP;
  payshap_id: string;
  reference_prefix?: string;
}

export interface ZapperPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.ZAPPER;
  merchant_id: string;
  store_id?: string;
  qr_code_url?: string;
}

export interface CardPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.CARD_PAYMENTS;
  merchant_id: string;
}

export interface MobileMoneyPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.MOBILE_MONEY;
  provider: string;
  merchant_code: string;
}

export interface CashPaymentInfo extends BasePaymentMethodInfo {
  payment_method: AcceptedPaymentMethods.CASH;
  // No additional fields needed for cash
}

// Union type for all payment methods
export type PaymentMethodInfo = 
  | EFTPaymentInfo 
  | InstantEFTPaymentInfo 
  | SnapScanPaymentInfo 
  | PayShapPaymentInfo 
  | ZapperPaymentInfo 
  | CardPaymentInfo 
  | MobileMoneyPaymentInfo 
  | CashPaymentInfo;

File: invoice_me/backend/models/payment_information.py
from pydantic import BaseModel, Field
from typing import Optional, Union, Literal
from models.invoices import AcceptedPaymentMethods
from enum import Enum

class BasePaymentMethodInfo(BaseModel):
    payment_method: AcceptedPaymentMethods
    enabled: bool = True
    information_present: bool = False
    display_name: Optional[str] = None
    description: Optional[str] = None

class EFTPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.EFT] = AcceptedPaymentMethods.EFT
    bank_name: str
    account_holder: str
    account_number: str
    branch_code: str
    swift_code: Optional[str] = None

class InstantEFTPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.INSTANT_EFT] = AcceptedPaymentMethods.INSTANT_EFT
    bank_name: str
    account_holder: str
    account_number: str
    branch_code: str

class SnapScanPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.SNAPSCAN] = AcceptedPaymentMethods.SNAPSCAN
    merchant_id: str
    store_id: Optional[str] = None
    qr_code_url: Optional[str] = None

class PayShapPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.PAYSHAP] = AcceptedPaymentMethods.PAYSHAP
    payshap_id: str
    reference_prefix: Optional[str] = None

class ZapperPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.ZAPPER] = AcceptedPaymentMethods.ZAPPER
    merchant_id: str
    store_id: Optional[str] = None
    qr_code_url: Optional[str] = None

class CardPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.CARD_PAYMENTS] = AcceptedPaymentMethods.CARD_PAYMENTS
    merchant_id: str

class MobileMoneyPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.MOBILE_MONEY] = AcceptedPaymentMethods.MOBILE_MONEY
    provider: str
    merchant_code: str

class CashPaymentInfo(BasePaymentMethodInfo):
    payment_method: Literal[AcceptedPaymentMethods.CASH] = AcceptedPaymentMethods.CASH

# This is the Union type port
PaymentMethodInfo = Union[
    EFTPaymentInfo,
    InstantEFTPaymentInfo,
    SnapScanPaymentInfo,
    PayShapPaymentInfo,
    ZapperPaymentInfo,
    CardPaymentInfo,
    MobileMoneyPaymentInfo,
    CashPaymentInfo
]

# We use Field(discriminator=...) so Pydantic knows which model to use 
# when parsing a list of mixed payment methods.
class AnnotatedPaymentMethod(BaseModel):
    info: Union[
        EFTPaymentInfo,
        InstantEFTPaymentInfo,
        SnapScanPaymentInfo,
        PayShapPaymentInfo,
        ZapperPaymentInfo,
        CardPaymentInfo,
        MobileMoneyPaymentInfo,
        CashPaymentInfo
    ] = Field(..., discriminator='payment_method')